### PR TITLE
Add support for MaxRL  

### LINKED ISSUE #5025: Maximum Likelihood Reinforcement Learning
### Method description

Add Maximum Likelihood Reinforcement Learning (MaxRL) as a new training objective for reinforcement learning tasks. MaxRL bridges traditional RL and maximum likelihood training by using p-normalization to concentrate learning signal on harder examples, achieving up to 20× test-time scaling efficiency gains.

## Motivation
Traditional RL (REINFORCE, GRPO) produces vanishing gradients on hard, low-success-rate problems
GRPO normalizes by standard deviation, which creates suboptimal weighting that inverts for easy problems
Models exhibit "pass@k degradation" - losing solution diversity during training

## Why MaxRL:
Better scaling with compute: Pareto-dominates GRPO across all tested benchmarks
More resistant to overfitting: Maintains pass@k diversity during extended training
Theoretically principled: Approximates maximum likelihood training via truncated Maclaurin expansion
Simple to implement: Single-line change from GRPO advantage calculation
Paper Reference: [Maximum Likelihood Reinforcement Learning (arXiv:2602.02710)](https://arxiv.org/abs/2602.02710)

### Open source status

- [x] The method implementation is available
- [ ] The model weights are available
- [x] The training datasets are available

### Provide useful links for the implementation

https://github.com/tajwarfahim/maxrl

### PR DESCRIPTION
**Pull request by catherinelee274**
# What does this PR do?

Adds Maxrl which is a variant of grpo with p-normalization.
Fixes #5025

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Changes GRPO advantage normalization logic by adding a new `scale_rewards="mean"` mode, which can affect training dynamics and stability for users enabling it (guarded by config validation).
> 
> **Overview**
> Adds an *experimental* MaxRL-style advantage formulation to `GRPOTrainer` by introducing `scale_rewards="mean"`, which scales advantages by the grouped mean reward (instead of std) when using `multi_objective_aggregation="sum_then_normalize"`.
> 
> Enforces this constraint in `GRPOConfig.__post_init__`, updates the GRPO docs/paper index with the MaxRL reference and recommended config, adds a runnable `examples/scripts/maxrl.py`, and extends GRPO trainer tests to cover the new scaling option and a basic MaxRL training run.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 84b88b78a31ccc3ebc39cbd170831ac19a0ea6d7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by LeonEricsson** [comment]
Doesn't MaxRL reduce to simply changing the advantage normalization denominator from `std(r)` to `mean(r)`?

```
# GRPO
A_i = (r_i - mean(r)) / (std(r) + eps)

# MaxRL
A_i = (r_i - mean(r)) / (mean(r) + eps)
```

If so, this fits naturally as a flag in the existing GRPO trainer rather than a dedicated experimental module.

File: trl/trainer/grpo_trainer.py [hunk 1]
```diff
@@ -1751,13 +1751,24 @@ def _generate_and_score_completions(
                     std_rewards = rewards.std().expand_as(rewards)
                 else:  # doesn't occur during training, but could occur in eval when num_generations_eval=batch_size=1
                     std_rewards = torch.zeros_like(rewards)
+            elif self.scale_rewards == "mean":
+                # MaxRL (https://arxiv.org/abs/2602.02710): normalize by the group mean rather than the group std.
+                # A_i = (r_i - mean(r)) / (mean(r) + eps). For binary rewards this is equivalent to normalizing
+                # by the empirical success rate. std_rewards is set to zeros here only for logging purposes
+                # (is_std_zero tracking); the actual scaling uses mean_grouped_rewards directly.
+                std_rewards = torch.zeros_like(rewards)
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py]:
```suggestion
```

Since `std_rewards` isn't used. Consider moving it to the existing branch instead:

```python
if self.scale_rewards in ["group", "none", "mean"]:
```

This ensures `std_rewards` is still set to a meaningful value for logging. Might be worth adding a comment noting that `"mean"` doesn't use the std but we compute it anyway for observability.

File: trl/trainer/grpo_trainer.py [hunk 2]
```diff
@@ -1751,13 +1751,24 @@ def _generate_and_score_completions(
...
+                    f"Invalid value for scale_rewards: {self.scale_rewards}. Must be one of 'batch', 'group', 'mean', or 'none'."
                 )
 
             advantages = rewards - mean_grouped_rewards
-            if self.scale_rewards != "none":
+            if self.scale_rewards == "mean":
+                # Experimental: MaxRL advantage normalization — divide by group mean instead of group std.
+                # When all rollouts in a group fail (mean == 0), the numerator is also 0, so the result is 0
+                # regardless of eps; abs() keeps the denominator positive for mixed-sign reward spaces.
+                advantages = advantages / (mean_grouped_rewards.abs() + 1e-4)
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py]:
```suggestion
                advantages = advantages / (mean_grouped_rewards.abs() + 1e-4)
```

This comment is heavier than it needs to be. The only non-obvious part is the `abs()`, which seems to deviate from MaxRL? Could you point me to where this appears in the MaxRL reference implementation? I couldn't find it there.

**Review by LeonEricsson — COMMENTED** [review]
Would you mind writing a paper index section for MaxRL as well?

File: tests/test_grpo_trainer.py [hunk 3]
```diff
@@ -1391,6 +1391,197 @@ def test_training_scale_rewards(self, scale_rewards):
...
+            [
+                (1.0 - mean0) / (mean0 + eps),
+                (0.0 - mean0) / (mean0 + eps),
+                (0.0 - mean0) / (mean0 + eps),
+                (1.0 - mean1) / (mean1 + eps),
+                (1.0 - mean1) / (mean1 + eps),
+                (0.0 - mean1) / (mean1 + eps),
+            ]
+        )
+        torch.testing.assert_close(advantages, expected)
```
Comment (LeonEricsson) [inline tests/test_grpo_trainer.py]:
```suggestion
```

This is fine as an offline test for the PR but doesn't need to be a part of the test suite

File: tests/test_grpo_trainer.py [hunk 4]
```diff
@@ -1391,6 +1391,197 @@ def test_training_scale_rewards(self, scale_rewards):
             new_param = trainer.model.get_parameter(n)
             assert not torch.equal(param, new_param), f"Parameter {n} has not changed."
 
+    # ------------------------------------------------------------------
+    # MaxRL tests (scale_rewards="mean")
+    # ------------------------------------------------------------------
```
Comment (LeonEricsson) [inline tests/test_grpo_trainer.py]:
```suggestion
```

File: tests/test_grpo_trainer.py [hunk 5]
```diff
@@ -1391,6 +1391,197 @@ def test_training_scale_rewards(self, scale_rewards):
...
+            model="trl-internal-testing/tiny-Qwen2ForCausalLM-2.5",
+            reward_funcs="trl-internal-testing/tiny-Qwen2ForSequenceClassification-2.5",
+            args=training_args,
+            train_dataset=dataset,
+            eval_dataset=dataset,
+        )
+
+        trainer.train()
+
+        assert trainer.state.log_history[-1]["train_loss"] is not None
```
Comment (LeonEricsson) [inline tests/test_grpo_trainer.py]:
```suggestion
```

File: tests/test_grpo_trainer.py [hunk 6]
```diff
@@ -1391,6 +1391,197 @@ def test_training_scale_rewards(self, scale_rewards):
...
+
+        mean_grouped = rewards.view(-1, num_generations).mean(dim=1)
+        mean_grouped = mean_grouped.repeat_interleave(num_generations, dim=0)
+
+        eps = 1e-4
+        advantages = (rewards - mean_grouped) / (mean_grouped + eps)
+
+        # numerator is 0 for all, denominator is eps → advantages all 0
+        assert not torch.isnan(advantages).any(), "advantages must not contain NaN"
+        torch.testing.assert_close(advantages, torch.zeros(3))
```
Comment (LeonEricsson) [inline tests/test_grpo_trainer.py]:
```suggestion
```

same for this. These tests are tautological, they reimplement behavior inline rather than testing trl code.

File: tests/test_grpo_trainer.py [hunk 7]
```diff
@@ -1391,6 +1391,197 @@ def test_training_scale_rewards(self, scale_rewards):
...
+        )
+
+        previous_trainable_params = {n: param.clone() for n, param in trainer.model.named_parameters()}
+
+        trainer.train()
+
+        assert trainer.state.log_history[-1]["train_loss"] is not None
+        for n, param in previous_trainable_params.items():
+            new_param = trainer.model.get_parameter(n)
+            assert not torch.equal(param, new_param), f"Parameter {n} has not changed."
```
Comment (LeonEricsson) [inline tests/test_grpo_trainer.py]: this seems like a duplicate of test_maxrl_training since we're not actually verifying anything related to the multiple reward functions.

File: tests/test_grpo_trainer.py [hunk 8]
```diff
@@ -1391,6 +1391,197 @@ def test_training_scale_rewards(self, scale_rewards):
...
```
Comment (LeonEricsson) [inline tests/test_grpo_trainer.py]: similar comment to test_maxrl_training_multiple_reward_funcs. I also don't see why the combination of MaxRL and conversatinal training need a dedicated test, isn't this already covered by existing conversational tests?

File: tests/test_grpo_trainer.py [hunk 9]
```diff
@@ -1391,6 +1391,197 @@ def test_training_scale_rewards(self, scale_rewards):
...
+            reward_funcs="trl-internal-testing/tiny-Qwen2ForSequenceClassification-2.5",
+            args=training_args,
+            train_dataset=dataset,
+            peft_config=LoraConfig(task_type="CAUSAL_LM"),
+        )
+
+        trainer.train()
+
+        assert trainer.state.log_history[-1]["train_loss"] is not None
+        assert isinstance(trainer.model, PeftModel)
```
Comment (LeonEricsson) [inline tests/test_grpo_trainer.py]: same thoughts as test_maxrl_training_multiple_reward_funcs and test_maxrl_training_conversational

Comment (catherinelee274) [reply tests/test_grpo_trainer.py]: removed

File: tests/test_grpo_trainer.py (hunk 8 above)
Comment (catherinelee274) [reply tests/test_grpo_trainer.py]: removed. Yes perhaps maxrl is too small of a new change to warrant new tests.

File: tests/test_grpo_trainer.py (hunk 7 above)
Comment (catherinelee274) [reply tests/test_grpo_trainer.py]: Removed.

File: tests/test_grpo_trainer.py (hunk 6 above)
Comment (catherinelee274) [reply tests/test_grpo_trainer.py]: Removed.

File: tests/test_grpo_trainer.py (hunk 5 above)
Comment (catherinelee274) [reply tests/test_grpo_trainer.py]: Removed.

File: tests/test_grpo_trainer.py (hunk 4 above)
Comment (catherinelee274) [reply tests/test_grpo_trainer.py]: Removed

File: tests/test_grpo_trainer.py (hunk 3 above)
Comment (catherinelee274) [reply tests/test_grpo_trainer.py]: Removed.

File: trl/trainer/grpo_trainer.py [hunk 10]
```diff
@@ -1820,8 +1820,10 @@ def _generate_and_score_completions(
             rewards = (rewards_per_func * self.reward_weights.to(device).unsqueeze(0)).nansum(dim=1)
             mean_grouped_rewards = rewards.view(-1, num_generations).mean(dim=1)
             mean_grouped_rewards = mean_grouped_rewards.repeat_interleave(num_generations, dim=0)
-            if self.scale_rewards in ["group", "none"]:
-                # If self.scale_rewards = "none", we'll only use std_rewards to check for zero std for logging
+            if self.scale_rewards in ["group", "none", "mean"]:
+                # For "none" and "mean", std_rewards is not used for advantage scaling but is computed
+                # for observability (is_std_zero logging). "mean" uses the group mean as denominator
+                # instead (MaxRL, https://arxiv.org/abs/2602.02710).
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py]:
```suggestion
                # for observability (is_std_zero logging). "mean" uses the group mean as denominator.
```

leave paper references out of the trainer, its well documented in the config.

File: trl/trainer/grpo_trainer.py [hunk 11]
```diff
@@ -1835,11 +1837,14 @@ def _generate_and_score_completions(
...
             else:
                 raise ValueError(
-                    f"Invalid value for scale_rewards: {self.scale_rewards}. Must be one of 'batch', 'group', or 'none'."
+                    f"Invalid value for scale_rewards: {self.scale_rewards}. Must be one of 'batch', 'group', 'mean', or 'none'."
                 )
 
             advantages = rewards - mean_grouped_rewards
-            if self.scale_rewards != "none":
+            if self.scale_rewards == "mean":
+                # MaxRL: A_i = (r_i - mean(r)) / (mean(r) + eps)
```
Comment (LeonEricsson) [inline trl/trainer/grpo_trainer.py]:
```suggestion
```

redundant, code is self-explanatory

**Review by LeonEricsson — COMMENTED** [review]
final comments. then i'm satisfied. 
needs a maintainers approval before merging.

**Comment by qgallouedec** [comment]
Thanks @catherinelee274, and sorry for the long delay. MaxRL is a nice paper and we could support it. Thanks for taking @LeonEricsson's suggestion and landing it as a flag, that's the right shape, and I checked the one-liner against Algorithm 1 (`(r_j − r̂)/r̂`): it's faithful. A few things before merge.

**Please drop `examples/scripts/maxrl.py`.** It's 167 of the 235 added lines for a one-flag change, the `paper_index.md` snippet covers it. It also reimplements a `\boxed{}` extractor and a binary accuracy reward when [`trl.rewards.accuracy_reward`](https://github.com/huggingface/trl/blob/main/trl/rewards/accuracy_rewards.py) exists (see `examples/scripts/gspo.py`), and hardcodes `training_args.scale_rewards = "mean"` after parsing, overriding the user's CLI.

**Binary rewards only: this needs to be explicit.** `advantages / (mean_grouped_rewards + 1e-4)` is well-defined only where the paper derives it; with a reward model or any negative reward the group mean can sit near zero or below, and the advantage flips sign or blows up. The paper says so directly: *"Our work currently assumes a binary reward setting and does not extend directly to continuous or arbitrarily valued rewards."* Please state that in the `scale_rewards` docstring and `paper_index.md`. You can keep `test_maxrl_training` as is (it currently uses `tiny-Qwen2ForSequenceClassification-2.5`i.e. arbitrary-signed rewards) we could solve it later. No runtime check needed; docs plus a representative test is enough imo.

**Keep the `sum_then_normalize` gate, but explain it.** It's correct, not incidental: under `normalize_then_sum` each reward function is z-scored per group before summing, so `rewards` is already ~zero-mean per group and dividing by `mean(r)` would be degenerate. Put that in a comment and in the error message, or the next reader will assume it's arbitrary and remove it.

**Question on `num_generations`.** MaxRL is compute-indexed (`N` sets the truncation level), and the paper prescribes no minimum `N` and no clipping on `1/r̂`. At TRL's typical `num_generations=8`, `r̂` is noisy and `1/r̂` has high variance. Did you look at small-`N` behaviour? If the method wants a large `num_generations`, that deserves a sentence in the docs.

**Rebase, and make the `mean` branch nan-aware.** Not just a textual merge: the advantage block now uses `torch.nanmean`/`nanstd` and has an `unscorable_mask` path for completions where every reward function returned `None`.

**Propagate to the duplicated copies.** The same `scale_rewards` block lives in `trl/experimental/dppo/dppo_trainer.py` and `trl/experimental/grpo_with_replay_buffer/grpo_with_replay_buffer_trainer.py`; those have to stay aligned (including the updated `ValueError`).

**Nits:** `"mean"` sits on a different axis from `"group"`/`"batch"` (statistic vs. scope). `"group_mean"` reads better? but happy to hear if you disagree. `test_maxrl_training` duplicates the parametrized `scale_rewards="mean"` case. `grpo_config.py` (docstring and `help`) links to `arxiv.org/abs/2602.02710`; we use `https://huggingface.co/papers/2602.02710`.

Thanks again, the core change is small and well-targeted, and lands easily once the script is gone and the binary-reward scope is nailed down.

**Comment by qgallouedec** [comment]
Closing this one, we're not taking MaxRL.

The demand isn't there. #5025 is the only request for it, you opened it yourself, and in 7 months it got no comment and no reaction from anyone else. The paper hasn't picked up traction either. Every flag we add to `GRPOConfig` is permanent surface, so the bar is a method people actually ask for.

Nothing wrong with the implementation, the one-liner was faithful to Algorithm 1. It's the method we're passing on, not your work. If MaxRL gets adopted later we can revisit.
